jupyter_sphinx/ast.py

```diff
@@ -358,3 +400,50 @@ def get_widgets(notebook):
         # Don't catch KeyError, as it's a bug if 'widgets' does
         # not contain 'WIDGET_STATE_MIMETYPE'
         return None
+
+
+class CellOutputsToNodes(SphinxTransform):
+    """Use the builder context to transform a CellOutputNode into Sphinx nodes."""
+
+    default_priority = 700
+
+    def apply(self):
+        thebe_config = self.config.jupyter_sphinx_thebelab_config
+
+        for cell_node in self.document.traverse(JupyterCellNode):
+            output_bundle_node = _return_first_node_type(
+                cell_node, CellOutputBundleNode
+            )
+            # Create doctree nodes for cell outputs.
+            output_nodes = cell_output_to_nodes(
+                output_bundle_node.outputs,
+                self.config.jupyter_execute_data_priority,
+                bool(cell_node.attributes["stderr"]),
+                sphinx_abs_dir(self.env),
+                thebe_config,
+            )
+            # Remove the outputbundlenode and we'll attach the outputs next
+            attach_outputs(output_nodes, cell_node, thebe_config, cell_node.cm_language)
+
+        # Image collect extra nodes from cell outputs that we need to process
+        for node in self.document.traverse(image):
+            # If the image node has `candidates` then it's already been processed
+            # as in-line markdown, so skip it
+            if "candidates" in node:
+                continue
+            col = ImageCollector()
```

Comment:
I'm a bit concerned that we might be not respecting the app settings here; could we reuse the image collector from the app somehow? I imagine different builders will have different image collectors.

Reply:
hmm that is a good question. It hasn't been an issue in the Jupyter Book project so far (and we've built both PDF as well as HTML outputs), though I agree it'd be better if we could reuse the image collector that the app uses

Reply:
I'm also OK leaving this as is and waiting for someone to report a problem.

Please add an extra comment why this is necessary though.